### ISSUE TITLE
Bias dimension not correctly updated when converting `nn.Conv1d` layers to `CONV2D`

### ISSUE THREAD

**Issue by etrommer** [google-ai-edge/litert-torch#674]
### Description of the bug:

When using `nn.Conv1d` layers, they are converted to TFlite `CONV2D` operators. However, the bias term does seem to be ignored in the process, leading to several additional `TRANSPOSE` operations in the output graph. When expressing the same functionality using an equivalent `nn.Conv2d` instance with a `(1xk)` kernel size, the bias is correctly added to the `CONV2D` operator in the TFlite graph.

Consider these two functionally equivalent models:
```python
class TinyConv1D(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = torch.nn.Conv1d(2, 4, kernel_size=3, bias=True)
        self.conv2 = torch.nn.Conv1d(4, 1, kernel_size=3)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        x = self.conv1(x)
        x = self.conv2(x)
        return x


class TinyConv2D(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = torch.nn.Conv2d(2, 4, kernel_size=(1, 3), bias=True)
        self.conv2 = torch.nn.Conv2d(4, 1, kernel_size=(1, 3))

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        x = self.conv1(x)
        x = self.conv2(x)
        return x
```
converted like so:
```python
def aiet_export(model: torch.nn.Module, fname: str, input_size: Tuple[int, ...]):
    sample_input = torch.rand(input_size)
    model = ai_edge_torch.to_channel_last_io(model, args=[0], outputs=[0]).eval()
    model(sample_input)
    model = export_for_training(model, (sample_input,)).module()
    aiet_tc = ai_edge_torch.convert(model, (sample_input,))
    aiet_tc.export(fname)

tc1d = TinyConv1D()
aiet_export(tc1d, "tinyconv1d.tflite", (1, 32, 2))

tc2d = TinyConv2D()
aiet_export(tc2d, "tinyconv2d.tflite", (1, 1, 32, 2))
```

### Expected
Equivalent output for TinyConv1D and TinyConv2D model definitions

### Actual
TinyConv1D contains several additional transposes and a separate bias addition with 3D dimensions that is not fused into the `CONV2D` operator:

![Image](https://github.com/user-attachments/assets/c0b16724-0b53-4089-a2b6-5d5d5d445e85)

TinyConv2D has a correctly transposed bias that is added to the `CONV2D` operator:

![Image](https://github.com/user-attachments/assets/d7084295-1b45-4ae5-908b-46a2149afd77)

### Any other information you'd like to share?

Environment:
- Python: 3.10.15
- Tensorflow: 2.19.0
- AI Edge Torch: 0.4.0
- PyTorch: 2.6.0

**Comment by pkgoogle** [comment]
Hi @dpknag,, can you please take a look? Thanks.

**Comment by dpknag** [comment]
Hi @etrommer,

Thanks for this detailed issue. We took a look at the TinyConv1D and TinyConv2D model conversions you outlined, specifically checking how well each model converted from PyTorch to its TFLite version. Here's what we found for the Mean Squared Error (MSE) between the PyTorch and TFLite outputs. These values are reproducible when using torch.rand for inputs with global seed 42.

TinyConv1D (PyTorch) vs. tinyconv1d.tflite: MSE ≈ 0.0380
TinyConv2D (PyTorch) vs. tinyconv2d.tflite: MSE ≈ 0.0237

These numbers give us a baseline for the numerical closeness achieved by the ai_edge_torch converter for each model type, showing some minor differences are introduced during their respective conversions. If you have any particular data you'd like us to try out with these models, please let us know.

On a related note, as you might be aware, TFLite currently doesn't have a dedicated, native Conv1d operator; it only provides Conv2d. This means nn.Conv1d layers are translated, which could be a factor. We'd be open to discussing or even helping raise a feature request for a native Conv1d in TFLite if you think that would be broadly beneficial.

Thanks!

**Comment by etrommer** [comment]
Hi @dpknag ,

thank you for the quick response! The concern in my case is not so much numerical accuracy, but rather inference time. The additional `TRANSPOSE` operators incur quite a bit of overhead and. In our specific case, we would like to run these models on the Ethos-U55 NPU, which does not support `TRANSPOSE` ops, so the additional transpositions mean that the model will need to go back and forth between CPU and NPU.

Of course, a workaround is to rewrite the models to use `nn.Conv2d` like above, so this is not a major issue, but it would be nicer if this could be handled automatically during the conversion.

I would assume that the reason that `Conv1d` is not an operator in TFlite is precisely that it can be expressed as a specialized `Conv2d`. I think this conversion is also generally working fine, the issue seems to be that the layer's bias is not _also_ converted to the appropriate shape matching the new `Conv2d` operation. As you can see in TinyConv1d, the addition of the bias still has shape `1 x 4 x 1`  (i.e. Conv1d with channel-first layout). If it could simply be transposed to `1 x 1 x 1 x 4` (i.e. Conv2d with channel-last layout) in the converter, it would probably be merged into the Conv2d operator correctly.

**Comment by dpknag** [comment]
Hi @etrommer, 

Thank you for the detailed explanation, really appreciate it. @pkgoogle could you please look at this issue further.

**Comment by pkgoogle** [comment]
Hi @etrommer, I'm guessing the seemingly roundabout conversion is due to the need to generally handle Conv1D conversions. My understanding is that Conv1D is quite prevalent with Audio models especially Audio -> other modality models. So my feeling is that a intrinsic Conv1D operator for LiteRT would be quite beneficial -- do you agree?

**Comment by etrommer** [comment]
@pkgoogle Sorry for the late response - yes, audio modeling is also the context in which we encountered this issue. Someone [already asked about Conv1D in the TF repo](https://github.com/tensorflow/tensorflow/issues/51692) and the response also indicates that converting to Conv2D is the intended behavior. My concern would be that upstreaming a new operator into TFlite would take a long time and would probably be a tough sell if it doesn't add additional functionality.

The issue seems to arise in TF, where the rewrite from Conv1D to Conv2D happens: https://github.com/tensorflow/tensorflow/blob/de18c68e0e063a9e877acc742d559d327f35ee87/tensorflow/compiler/mlir/lite/stablehlo/transforms/legalize_hlo_conversions/conv.cc#L647
Presumably the cleanest way to fix this would be to extend the rewrite rule in TF to also consider the bias if present.

An alternative might be to handle this conversion already during the optimization passes in FX, by adding an additional pass that first converts Conv1D to Conv2D, then handles the shape transposition like is already implemented:
https://github.com/google-ai-edge/ai-edge-torch/blob/fe1307cd7601acc7f60eb60862938e50b57f9322/ai_edge_torch/_convert/fx_passes/optimize_layout_transposes_pass/layout_rewrite.py#L235-L258C17
although I understand that that isn't a great solution since it duplicates code from TF and injects TFlite idiosyncrasies into a place where they don't really belong.

**Comment by pkgoogle** [comment]
Yeah that makes sense, we'll probably go for the 2nd one, and you are correct a new op might take awhile.
